### ISSUE TITLE
SSO Example fails with UnhandledPromiseRejection

### ISSUE THREAD

**Issue by ZIRaja** [OfficeDev/generator-office#539]
Hello.

I used the following [tutorial](https://docs.microsoft.com/en-us/office/dev/add-ins/quickstarts/sso-quickstart) to create a Outlook-Addin with SSO and Microsoft Graph support enabled.

However, like it is also mentioned on the page, I configured the Addin using [this tutorial](https://docs.microsoft.com/de-de/office/dev/add-ins/develop/create-sso-office-add-ins-nodejs#register-the-add-in-with-azure-ad-v20-endpoint), creating the entry in Azure AD and everything.

After doing all these steps and running npm, I get the following error message:

>[0]
[0] > office-addin-taskpane-sso@0.0.0 start:server C:\Desktop\Addin
[0] > office-addin-sso start manifest.xml
[0]
[0] (node:10640) UnhandledPromiseRejectionWarning: false
[0] (node:10640) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
[0] (node:10640) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
[0] npm run start:server exited with code 0

I have followed the tutorial exactly as described, but still the error remains. I am using:

- Windows 10 Build 18363
- Node v12.16.3
- Yeoman v3.1.1

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 3020ff5d-ee3b-976b-f1ad-0ed378fdb47b
* Version Independent ID: 5de0010a-6435-154b-a7fd-fd2ea05ebc0b
* Content: [Use the Yeoman generator to create an Office Add-in that uses SSO (preview) - Office Add-ins](https://docs.microsoft.com/en-us/office/dev/add-ins/quickstarts/sso-quickstart#configure-sso)
* Content Source: [docs/quickstarts/sso-quickstart.md](https://github.com/OfficeDev/office-js-docs-pr/blob/master/docs/quickstarts/sso-quickstart.md)
* Product: **non-product-specific**
* Technology: **add-ins**
* GitHub Login: @o365devx
* Microsoft Alias: **o365devx**

**Comment by TCourtneyOwen** [comment]
@ZIRaja Sorry to hear you are having problems.  I assume you ran the "npm run configure-sso" step and that this succeeded without errors before running npm start? I will try this scenario out today and see if I can repro the problem.  I will get back to you.

Thanks,

Courtney

**Comment by TCourtneyOwen** [comment]
@ZIRaja Also, can you please clarify if you are using the TypeScript or JavaScript SSO template?

Thanks,

Courtney

**Comment by ZIRaja** [comment]
I used the TypeScript SSO template.
I ran into issues running configure-sso (probably because of the two-factor authentication and missing admin privileges), so I had to follow the second tutorial where I registered the application on Azure AD and then added those values to the appropriate files as described.

**Comment by ZIRaja** [comment]
Yeah, I followed those steps too. My ENV file contains CLIENT_ID, CLIENT_SECRET, GRAPH_URL_SEGMENT, NODE_ENV, PORT, and SCOPE, all set to the values from the AD registration.
However, the manual does not mention the Credential Manager anywhere. Additionally, there is no (straightforward) way to add Web Credentials in Windows 10 apparently.

**Comment by TCourtneyOwen** [comment]
@ZIRaja Were you ever able to resolve this issue?  You shouldn't need to make any manual changes to the Credential Manager.  This should all be done automatically when you run "npm run configure-sso" after creating your sso project

**Comment by pauljereb** [comment]
We have the same issue when the add-in is started on a different machine than where "npm run configure-sso" was run.

**Comment by TCourtneyOwen** [comment]
@pauljereb I think the problem you are running into is that you can't simply move the project that you ran "configure-sso" on over to another machine without also updating the secret that was generated during the "configure-sso" process over to the Credential Manager on the other machine.  The secret is outputted to the console when running "configure-sso".  You would need to copy that secret and then move it over to the other machine and put it in Windows Credential Manager --> Web Credentials.  The secret name The name that use in Web Credentials should be the same name that you provided for the project when you created it.  Do you happen to remember what the secret was that generated during configure-sso?  This secret is need to be able to connect the to Azure application that was created when running configure-sso.

-Courtney

**Comment by pauljereb** [comment]
thanks for your explanation @TCourtneyOwen. for being able to run the project in Linux, or for easier handling in development teams, it would be beneficial to have an option in "configure-sso" to choose between Windows Credential Manager and, for example, storing the Client Secret in the .env file (for development).

**Comment by arozwalak** [comment]
I'm having the same issue as above. I've installed yeoman generator-office and run yo office, to create add-in for outlook with sso. then I've run npm run configure-sso, which finished successfully. But then when I run 'npm start' I've got

[0] (node:5207) UnhandledPromiseRejectionWarning: false
[0] (node:5207) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
[0] (node:5207) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

**Comment by arozwalak** [comment]
I just found that it might be related to add-in name provided during configuration. When I provided name with spaces (as proposed on microsoft documentation) I'm getting the above problem. but when I used '-' instead of spaces, then it worked. Hope that helps

**Comment by TCourtneyOwen** [comment]
@arozwalak OK good to know, let me see if I can repro that and if so I will try and get a fix out in the next day or so.

-Courtney

**Comment by TCourtneyOwen** [comment]
@arozwalak I just tried out creating a JavaScript Outlook SSO project with a space in the project name (" sso test"), and didn't run into any issues.  I am running Node v12.18.4

Is there any additional error information beyond this:

[0] (node:5207) UnhandledPromiseRejectionWarning: false
[0] (node:5207) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
[0] (node:5207) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

Thanks,

Courtney

**Comment by arozwalak** [comment]
On the opposite site. I tried to run application today and I've got the same error (yesterday it was running without any issues, and I didn't made any changes to add-in). below is full output from `npm start` command

`> office-addin-taskpane-sso@0.0.0 start /Users/artur/workspace/vic-outlook-add-in
> npm run build:dev && concurrently "npm run start:server"


> office-addin-taskpane-sso@0.0.0 build:dev /Users/artur/workspace/vic-outlook-add-in
> webpack --mode development --https false

Hash: b2983ce9e26d83ff391d
Version: webpack 4.43.0
Time: 5522ms
Built at: 10/09/2020 11:36:35 AM
                    Asset       Size              Chunks                   Chunk Names
       assets/icon-16.png   1.56 KiB                      [emitted]
       assets/icon-32.png   2.33 KiB                      [emitted]
       assets/icon-80.png   4.72 KiB                      [emitted]
   assets/logo-filled.png   11.6 KiB                      [emitted]
            commands.html  471 bytes                      [emitted]
              commands.js   5.89 KiB            commands  [emitted]        commands
          commands.js.map   6.04 KiB            commands  [emitted] [dev]  commands
  fallbackauthdialog.html  801 bytes                      [emitted]
    fallbackauthdialog.js    354 KiB  fallbackauthdialog  [emitted]        fallbackauthdialog
fallbackauthdialog.js.map    340 KiB  fallbackauthdialog  [emitted] [dev]  fallbackauthdialog
         manifest.dev.xml   4.46 KiB                      [emitted]
              polyfill.js    402 KiB            polyfill  [emitted]        polyfill
          polyfill.js.map    322 KiB            polyfill  [emitted] [dev]  polyfill
             taskpane.css    1.7 KiB                      [emitted]
            taskpane.html   4.01 KiB                      [emitted]
              taskpane.js    616 KiB            taskpane  [emitted]        taskpane
          taskpane.js.map    733 KiB            taskpane  [emitted] [dev]  taskpane
Entrypoint commands = commands.js commands.js.map
Entrypoint fallbackauthdialog = fallbackauthdialog.js fallbackauthdialog.js.map
Entrypoint polyfill = polyfill.js polyfill.js.map
Entrypoint taskpane = taskpane.js taskpane.js.map
[0] util (ignored) 15 bytes {taskpane} [built]
[1] util (ignored) 15 bytes {taskpane} [built]
[./assets/icon-16.png] 64 bytes {taskpane} [built]
[./assets/icon-32.png] 64 bytes {taskpane} [built]
[./assets/icon-80.png] 64 bytes {taskpane} [built]
[./node_modules/webpack/buildin/global.js] (webpack)/buildin/global.js 472 bytes {commands} {taskpane} [built]
[./node_modules/webpack/buildin/module.js] (webpack)/buildin/module.js 497 bytes {taskpane} [built]
[./src/commands/commands.ts] 1.15 KiB {commands} [built]
[./src/helpers/fallbackAuthHelper.ts] 5.31 KiB {taskpane} [built]
[./src/helpers/fallbackauthdialog.ts] 2.54 KiB {fallbackauthdialog} [built]
[./src/helpers/ssoauthhelper.ts] 8.56 KiB {taskpane} [built]
[./src/taskpane/taskpane.ts] 1.73 KiB {taskpane} [built]
    + 418 hidden modules
Child HtmlWebpackCompiler:
                     Asset      Size  Chunks             Chunk Names
    assets/logo-filled.png  11.6 KiB          [emitted]
     + 3 hidden assets
    Entrypoint HtmlWebpackPlugin_0 = __child-HtmlWebpackPlugin_0
    Entrypoint HtmlWebpackPlugin_1 = __child-HtmlWebpackPlugin_1
    Entrypoint HtmlWebpackPlugin_2 = __child-HtmlWebpackPlugin_2
    [./assets/logo-filled.png] 68 bytes {HtmlWebpackPlugin_0} [built]
    [./node_modules/html-webpack-plugin/lib/loader.js!./src/commands/commands.html] 448 bytes {HtmlWebpackPlugin_1} [built]
    [./node_modules/html-webpack-plugin/lib/loader.js!./src/helpers/fallbackauthdialog.html] 779 bytes {HtmlWebpackPlugin_2} [built]
    [./node_modules/html-webpack-plugin/lib/loader.js!./src/taskpane/taskpane.html] 4.17 KiB {HtmlWebpackPlugin_0} [built]
[0]
[0] > office-addin-taskpane-sso@0.0.0 start:server /Users/artur/workspace/vic-outlook-add-in
[0] > office-addin-sso start manifest.xml
[0]
Password:
[0] (node:49231) UnhandledPromiseRejectionWarning: false
[0] (node:49231) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
[0] (node:49231) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.`


this was quite strange, so I stopped and run application again couple of time and finally it started without problems, so looks like this issue is random.

**Comment by jasonjoh** [comment]
I reproduced this problem trying to not use the "configure-sso" option. I think that the instructions on how to register and configure the app are missing some steps. It seems to hint that configure-sso isn't required (since it doesn't work in some cases) and that you can manually configure. Do you really have to add the secret in Credentials Manager? Does this work on Mac/Linux?

**Comment by TCourtneyOwen** [comment]
You definitely need to run the "configure-sso" step.  Which instructions are you following?  After creating the SSO project with Yo Office, you should see something like the following in the output:

      Congratulations! Your add-in has been created! Your next steps:

      1. Go the directory where your project was created:

         cd C:\ssoTest

      2. Configure your SSO taskpane add-in:

         npm run configure-sso

      3. Start the local web server and sideload the add-in:

         npm start

      4. Open the project in VS Code:

         code .

**Comment by jasonjoh** [comment]
Doing the SSO quickstart here: https://docs.microsoft.com/en-us/office/dev/add-ins/quickstarts/sso-quickstart#configure-sso

Specifically, it says:

> This command (meaning `npm run configure-sso`) will fail if your tenant is configured to require two-factor authentication. In this scenario, you'll need to manually complete the Azure app registration and SSO configuration steps, as described in the [Create a Node.js Office Add-in that uses single sign-on tutorial](https://docs.microsoft.com/en-us/office/dev/add-ins/develop/create-sso-office-add-ins-nodejs).

This doesn't tell you what to do with the secret. I was able to run configure-sso so I'm past that. Now I just need to figure out why Excel isn't opening when I run the project :D

**Comment by arozwalak** [comment]
I've cloned my addon to show as an example https://github.com/arozwalak/office-addon

when I run `npm start` I'm getting UnhandledPromiseRejectionWarning

**Comment by TCourtneyOwen** [comment]
@arozwalak This is very helpful. I will have a look at your project today and let you know what I find.

Thanks,

Courtney

**Comment by TCourtneyOwen** [comment]
I just cloned your project, added the required .ENV file at root (it was missing( created a new azure application via "configure-sso" and ran "npm start" without errors.  

You will need to have a .ENV file present at the root of your project as office-addin-sso reads from that.  Can you please add a .ENV file at the root of the project and update the value and let me know what you find?

CLIENT_ID=**your-application-id**
GRAPH_URL_SEGMENT=/me
NODE_ENV=development
PORT=3000
QUERY_PARAM_SEGMENT=
SCOPE=User.Read

Thanks,

Courtney

**Comment by georgeevans1995** [comment]
> I just found that it might be related to add-in name provided during configuration. When I provided name with spaces (as proposed on microsoft documentation) I'm getting the above problem. but when I used '-' instead of spaces, then it worked. Hope that helps

This is still an issue for mac, its happening in the getSecretFromCredentialStore function:


```
export function getSecretFromCredentialStore(ssoAppName: string, isTest: boolean = false): string {
    try {
        switch (process.platform) {
            case "win32":
                const getSecretFromWindowsStoreCommand = `powershell -ExecutionPolicy Bypass -File "${defaults.getSecretCommandPath}" "${ssoAppName}" "${os.userInfo().username}"`;
                return execSync(getSecretFromWindowsStoreCommand, { stdio: "pipe" }).toString();
            case "darwin":
                const getSecretFromMacStoreCommand = `${isTest ? "" : "sudo"} security find-generic-password -a ${os.userInfo().username} -s ${ssoAppName} -w`;
                return execSync(getSecretFromMacStoreCommand, { stdio: "pipe" }).toString();;
            default:
                const errorMessage: string = `Platform not supported: ${process.platform}`;
                throw new Error(errorMessage);
        }
    } catch {
        return '';
    }
}
```
In the above code the win32 case wraps ${ssoAppName} in double quotes and in the darwin case it doesn't. Once you add the double quotes it works as expected.

**Comment by andrewconnell** [comment]
@jasonjoh did you figure this out? I too am having the same problem... on macOS, not running `configure-sso` & registering the app manually. Can't figure out where the secret should go... 

Can't figure out what `configure-sso` does on the host machine, or if it's required... appears it just runs the Azure CLI [based on the docs](https://docs.microsoft.com/en-us/office/dev/add-ins/quickstarts/sso-quickstart#configure-sso)... 

Every time I run `npm start` I get the same `(node:17041) UnhandledPromiseRejectionWarning: false` errors others are getting & the process eventually bombs out before I can test the addin.

**Comment by TCourtneyOwen** [comment]
@andrewconnell The secret needs to go in the Keychain on Mac.  You can try running this command, adding in your own values to set the secret in the Keychain:

sudo security add-generic-password -a <user_name> -U -s <sso_project_name> -w <secret>

For the <sso_project_name> param this should be the DisplayName set in the SSO project manifest

**Comment by andrewconnell** [comment]
@TCourtneyOwen I'm confused... I've been pouring over the docs and can't seem to figure this out...

1. if I'm putting this in the keychain, then what's the process for deploying this add-in to be hosted elsewhere... where is the secret pulled from? Why not `.env`? I got the project working from the cloned sample, but it's so very different from the project that's created by the generator, it's hard to understand how the "black box" the generator creates works... 
2. What's the `<user_name>` value in the command you provide... my macOS login or the login of the AAD account in my demo tenant?

**Comment by TCourtneyOwen** [comment]
We opted to not include the ability of read the secret from a .env file because of security implications (e.g. someone puts there secret in the .env file and checks this into their repo). We didn't want to steer people down the path of where they might compromise the security of their secret.  However, if you want to do this, you could make a minor change to office-addin-sso to read the secret. You can see a prototype of this change in https://github.com/TCourtneyOwen/Office-Addin-Scripts/tree/add-ability-read-secret-from-env-file, specifically in the getSecret function in packages/office-addin-sso/src/server.ts

**Comment by TCourtneyOwen** [comment]
We realize this is shortcoming of the office-addin-sso package, but security ultimately trumped including the ability to read the secret from an .env by default.  I believe we are going to write documentation that tells people how they can port their SSO project over from localhost to another environment, but that's not yet available. Looping in @davidchesnut  to confirm.

**Comment by TCourtneyOwen** [comment]
Looping in @millerds as well, as he is new to the team and needs to be aware of this issue

**Comment by andrewconnell** [comment]
Thanks for the update, but the contradiction in the docs makes this very confusing.

When the tooling (_that's used for the [SSO Quick Start](https://docs.microsoft.com/en-us/office/dev/add-ins/quickstarts/sso-quickstart))_ forces you down this path, but the steps to _[Create an SSO addin with Node.js](https://docs.microsoft.com/en-us/office/dev/add-ins/develop/create-sso-office-add-ins-nodejs)_ don't use the tooling & an entirely different approach (_clone a repo that doesn't use any of the util scripts_)... 

Speaking as someone who writes a lot of docs for MSFT & MS Learning... this is hard to follow & is confusing.

**Comment by TCourtneyOwen** [comment]
I concur, this is confusing. The SSO Quick Start should have steps for deploying a locally hosted project to another environment (e.g. a production environment)

**Comment by Rick-Kirkham** [comment]
As I unserstand it, the QuickStart and the tooling are for people who want to treat SSO as a black box; but the "Create ..." article is intended to help people understand SSO and what's going on under the hood.

**Comment by andrewconnell** [comment]
@Rick-Kirkham said

> As I unserstand it, the QuickStart and the tooling are for people who want to treat SSO as a black box; but the "Create ..." article is intended to help people understand SSO and what's going on under the hood.

Not to be too picky, but...

IMHO: if the "Create..." isn't intended to be a black box, it shouldn't be using all these special `office-addin-*` util scripts. The QuickStart is MUCH easier to follow as it's a real Node.js project... not something with these baked runners/hosts.

I get what you guys are trying to do with those scripts, but the more abstraction, the harder it is to understand what is going on... frankly... the way it is now is backward.

_Just my $0.02_

**Comment by Rick-Kirkham** [comment]
I think there is a confusion here. The article https://docs.microsoft.com/en-us/office/dev/add-ins/develop/create-sso-office-add-ins-nodejs (The "Create ..." article) does **not** use `office-addin-*` scripts, but the QuickStart **does**.

**Comment by andrewconnell** [comment]
I stand corrected :)

I meant to say that the article that walks through "Create an SSO addin using Node.js" shouldn't start with "[Clone the starter project here](https://docs.microsoft.com/en-us/office/dev/add-ins/develop/create-sso-office-add-ins-nodejs#set-up-the-starter-project)"

**Comment by andrewconnell** [comment]
BTW, +💯 to https://github.com/OfficeDev/generator-office/issues/539#issuecomment-741939945... that's the issue on macOS & Node v14.

Fixed my issue #614. Hope to see his PR https://github.com/OfficeDev/Office-Addin-Scripts/pull/351 merged & released. Until then, developers on macOS need to patch the script themselves or just not create addins with spaces in their names.

**Comment by Rick-Kirkham** [comment]
This is a common way to create a tutorial or walkthrough about subject "X". We don't want the reader to spend time coding up the parts that aren't relevant to "X", so we have them start with a project where all the non-X parts are already done. 

The purpose of a quickstart is just to validate that the user has their development environment set up correctly. It's not supposed to teach them anything about the platform. 

Maybe the "Create ... " article should end with a note saying something like 

> "Now that you know how to code SSO in a Node-based add-in, in the future you should begin SSO-enabled Node-based add-ins with the Yo Office tool ... etc. ... The tool will install a library called office-addins-sso that contains code similar to what you wrote in this tutorial." 

In addition, the "Next Steps" at the end of the QuickStart should say "Optionally, if you want to understand how SSO is coded, go to "Create ... " article."

**Comment by Chris2011** [comment]
In my case, everything went fine with configure-sso on Windows, but this is not the live app later. My company still has an app and an ID and a secret now I need to change the secret from my local windows credential manager app to the secret that is needed for the correct app, but I can't change the credentials anymore. I also can't create web credentials manually via GUI, will check the power shell for that.

**Comment by Chris2011** [comment]
So in general, if I create a new app, this is just working fine for windows, for me. Also as the others already said, there is no information about the windows credential manager. Only we can see it in this ticket. For now I added the code from november to my server.js file to read the sercret from the env and not from the store. Thx for pointing this out :)

**Comment by Chris2011** [comment]
Thx @andrewconnell. It looks much cleaner and better :). Thx a alot.

**Comment by jlara005** [comment]
Hi @andrewconnell,

I keep getting this same error still. I followed the learning module step-by-step and added the secret to my web credential store using the PowerShell script (made sure it was same name as the project/display name). I had to configure the project manually since the "npm run configure-sso" doesn't work for me after I sign-in to azure. Anything else I need to check that might cause this error?

**Comment by millerds** [comment]
This hasn't been active for months.  Closing . . . but if someone feels more needs to be said or done please re-open.

**Comment by Chris2011** [comment]
@Nathure thx for the interest. I did the same as @TCourtneyOwen mentioned. I looked at the repo and copied those lines into my server.ts file: https://github.com/TCourtneyOwen/Office-Addin-Scripts/blob/add-ability-read-secret-from-env-file/packages/office-addin-sso/src/server.ts#L45-L58

**Comment by Haaxor1689** [comment]
For anyone that wants to use `.env` file to store the CLIENT_SECRET and doesn't want to fork the package or run the whole configure-sso script, I made a short script that only does the secret adding into windows credentials part using the ps1 script from the library.

js```
require('dotenv').config();
const { execSync } = require('child_process');
const { userInfo } = require('os');

const path = 'node_modules/office-addin-sso/lib/scripts/addAppSecret.ps1';
const appName = 'YOUR_APP_NAME';
const user = userInfo().username;
const secret = process.env.CLIENT_SECRET;

execSync(
	`powershell -ExecutionPolicy Bypass -File "${path}" "${appName}" "${user}" "${secret}"`,
	{ stdio: 'pipe' }
);
```

Just drop this script into the root of your project and anyone that will clone the project fresh can simply run it with a command like this:

```json
"secret:add": "node ./addAppSecret.js"
```

**Comment by SherylDarroch365** [comment]
@andrewconnell - I am creating an Excel add-in with Node JS and SSO, and my tenant uses MFA. I have been fighting with this for days trying to follow all of Microsoft's documents, to no avail. I finally came across this post and followed your tutorial including the PowerShell to add the secret to Windows Credential Manager, but I am still getting the "Credential store does not have secret for 'MyAddIn'" error. Any ideas? 

I have all the latest versions of the tech stack installed. I'm running out of time to deliver this add-in to my client. :(

**Comment by davidchesnut** [comment]
Hi @SherylDarroch365,

You might find it easier to work with the [Office Add-in Node.js SSO sample](https://github.com/OfficeDev/Office-Add-in-samples/tree/main/Samples/auth/Office-Add-in-NodeJS-SSO). It stores the secret in the .ENV file, although you will need to manually create the app registration. .ENV is fine for testing, but If you deploy to something like an Azure app service for production, you should move the secret to something like Azure key vault. I believe you would need to [update this sample file](https://github.com/OfficeDev/Office-Add-in-samples/blob/main/Samples/auth/Office-Add-in-NodeJS-SSO/Complete/server-helpers/obo-auth-helper.js) to retrieve the secret from key vault instead of .ENV.

Also the SSO template isn't really intended for use in production environments. Another reason I'd recommend starting with the sample, unless you are just experimenting with or studying SSO.

Hope this helps,
David

**Comment by SherylDarroch365** [comment]
@davidchesnut 

Thanks for the quick response. Your suggestions have been very helpful.

**Comment by SherylDarroch365** [comment]
Hi @davidchesnut,

I created my add-in starting from the Office Add-in Node.js SSO sample as you suggested, and now I am ready to deploy it to production. Unfortunately, when I use the command npm run build, I get a Module not found error, since there is no src folder in this project. 

I created a src folder and moved the pertinent files into that folder, and then upon build I received 32 errors that were mostly related to node modules not being found. The node modules folder was not moved into the src folder. 

Can you point me to instructions on how to get this project ready for production? 
Also, I understand that it has to be deployed as an API in Azure since it is not a static website, is that correct?

I appreciate any guidance you can provide.

Thanks,
Sheryl

**Comment by davidchesnut** [comment]
That sample doesn't use a build (webpack or other). Therefore no build is necessary. Just deploy the root folder.
Correct, you need to deploy to an Azure App Service app for the server side APIs to work.
 
I have a new article to publish soon that walks you through the steps to configure the Azure App Service (https://github.com/OfficeDev/office-js-docs-pr/pull/3790) It's based on the yo office SSO sample (with recent updates it does support deployment now.) Once that is published I plan to update the sample readme to include steps on how to deploy the sample based on that article. It is very similar, except you won't have a dist folder since there is no build. Instead you need to deploy the root folder of the project (views, public, all those folders). 

Hope this helps,
David

**Comment by davidchesnut** [comment]
@SherylDarroch365 the new publishing article is here: [Deploy a single sign-on (SSO) Office Add-in to Microsoft Azure App Service](https://learn.microsoft.com/office/dev/add-ins/publish/deploy-office-add-in-sso-to-azure). This is based on deploying a yo office generate project. Hope this helps!
